sympy/integrals/prde.py

```diff
@@ -441,6 +441,42 @@ def prde_no_cancel_b_small(b, Q, n, DE):
     return f + H, A.col_join(B).col_join(C)
 
 
+def prde_cancel_liouvillian(b, Q, n, DE):
+    """
+    Pg, 237.
+    Examples of this are:
+    log(x + 1)*log(x), log(x + 1)/log(x)
+
+    """
+    m = len(Q)
+    f, A = [None]*(n + 1), [None]*(n + 1)
+
+    # Why use DecrementLevel? Below line answers that:
+    # Assuming that we can solve such problems over 'k' (not k[t])
+    with DecrementLevel(DE):
+        DE0 = DE
```

Comment:
OK, I took a look at the code (I was on a plane when I wrote the above so I couldn't check). Perhaps the API should be modified so that DifferentialExtension is immutable. The API of risch is all internal, so there are no issues with changing it. In fact, I would highly expect it to need to change as progress is made in the algorithm.

So I suggest:

- Modify `decrement_level` and `increment_level` to return a new DifferentialExtension object. It should do so without redoing any unnecessary calculations. 
- `with DecrementLevel(DE)` would be changed to `with DecrementLevel(DE) as DE0`.  This is perhaps more annoying, as you have to use `DE0` inside the context manager instead of `DE`, but it would let you mix the original DE with the decremented one (which you apparently want to do here).